### PR TITLE
[FluentSystemIcon] Adding a new lib with all FluentUI icons

### PR DESCRIPTION
**Pull request by dvoituron**
# FluentSystemIcon

This new component will replace the FluentIcon (in the next PR), to have the possibility to use the **FluentUI.Icons** lib. Same for **FluentEmoji** component

> A new solution **src\Icons\FluentUI.Icons.sln** contains these new projects.

```razor
<FluentSystemIcon Icon="@(Icons.Regular.Size24.Save)" />
<FluentSystemIcon Icon="@Icons.Regular.Size48.Database" Color="Color.Custom" CustomColor="red" Width="12px;" />
<FluentSystemIcon Value="@(Icon.FromImageUrl("/Blazor.png"))" />

<FluentSystemEmoji Emoji="@(Emojis.PeopleBody.Color.Default.Artist)" />
```

I updated the Solution with sub-folder and 2 new projects: the **FluentUI.Icons lib** and the **Icons Generator** (see below).

image

---

# FluentUI Icons & Emoji's Generator

## How to generate FluentUI Icons classes

This application generates **FluentUI Icons** from SVG images
from the official website https://github.com/microsoft/fluentui-system-icons

To generate the icons, you need:

1. Download SVG images in a local folder (ex. C:\Temp) from the [official NPM repo](https://www.npmjs.com/package/@fluentui/svg-icons), using this command.

   ```cmd
   npm install @fluentui/svg-icons
   ```

2. a) Compile the project and run the `FluentUIIconsGenerator.exe` application,
   passing the **Assets** path to this local folder containing the SVG images as a parameter.
   
	   ```cmd
	   FluentUIIconsGenerator.exe --Assets=C:\Temp\Icons --Target=./Samples --Library=Icon
	   ```
   
   b) Or run the project, passing parameters to the `Properties/launchSettings.json` file:

	   ```json
	   {
		  "profiles": {
			"Microsoft.Fast.Components.FluentUI.IconsGenerator": {
			  "commandName": "Project",
			  "commandLineArgs": "--Assets=C:/Temp/Icons --Target=./Samples --Library=Icon"
			}
		  }
		}
	   ```

3. All generated C# files are located in the **Target** folder, or in the current folder if this attribute has not been defined.

## How to generate FluentUI Emoji classes

1. Download the [repository](https://github.com/microsoft/fluentui-emoji).

	```cmd
	git clone https://github.com/microsoft/fluentui-emoji
	```

2. a) Compile the project and run the `FluentUIIconsGenerator.exe` application,
   passing the **Assets** path to this local folder containing the SVG images as a parameter.
   
	   ```cmd
	   FluentUIIconsGenerator.exe --Assets=C:\Temp\Emojis --Target=./Samples --Library=Emoji
	   ```
   
   b) Or run the project, passing parameters to the `Properties/launchSettings.json` file:

	   ```json
	   {
		  "profiles": {
			"Microsoft.Fast.Components.FluentUI.IconsGenerator": {
			  "commandName": "Project",
			  "commandLineArgs": "--Assets=C:/Temp/Icons --Target=./Samples --Library=Emoji"
			}
		  }
		}
	   ```

## Help

Run the application with the `--help` parameter to see all available parameters.

```cmd
> FluentUIIconsGenerator.exe --help


  FluentIconsGenerator --folder:<Icons_Folder_Directory>

  --Assets    | -a   The root directory containing all SVG icons,
                     downloaded from https://github.com/microsoft/fluentui-system-icons.
                     If not specified, the current working directory will be used.

  --Library   | -l   The type of library to generate: icon or emoji.
                     If not specified, "icon" will be used.

  --Namespace | -ns  The namespace used for generated classes.
                     If not specified, "Microsoft.Fast.Components.FluentUI" will be used.

  --Names     | -n   The list of icon names to generate, separated by coma.
                     Example of icons: accessibility_32_filled,add_circle_20_filled
                     Example of emojis: accordion_flat,ambulance_high_contrast
                     By default: all icons

  --Sizes     | -s   The list of icon sizes to generate, separated by coma.
                     Example: 12,24. By default: all sizes
                     (Not available for emoji library)

  --Target    | -t   The target directory where C# classes will be created.
                     If not specified, the current working directory will be used.

  --Help      | -h   Display this documentation.
```

### CODE REVIEW

File: src/Microsoft.Fast.Components.FluentUI/Components/Icons/CoreIcons.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+//     This code was generated by a tool.
+//     Changes to this file may cause incorrect behavior and will be lost if
+//     the code is regenerated.
+// </auto-generated>
+
+#pragma warning disable 1591
+
+namespace Microsoft.Fast.Components.FluentUI;
+
+internal static partial class CoreIcons
```
Comment (vnbaaij) [inline src/Microsoft.Fast.Components.FluentUI/Components/Icons/CoreIcons.cs:11]: Do these icons exist twice (here and in lib)?

**Review by vnbaaij — COMMENTED** [review]
Impressive work!! Think we need to work on the readme's a bit more (but we can do so after merge). 
We probably do not want people to just install the Emoji/Icon packages, right? Do we need/want to add main/core package as a dependency/requirement?

**Comment by dvoituron** [comment]
Yes. Maybe the ReadMe files must be more detailed. 

It's not possible to install the Emoji or Icon lib without the Core lib. The reference us already there and I'm using the Icon/Emoji class defined in the Core lib.

File: src/Icons/Generator/ReadMe.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
+# FluentUI Icons & Emoji's Generator
+
+## How to generate FluentUI Icons classes
+
+This application generates **FluentUI Icons** from SVG images
+from the official website https://github.com/microsoft/fluentui-system-icons
+
+To generate the icons, you need:
+
+1. Download SVG images in a local folder (ex. C:\Temp) from the [official NPM repo](https://www.npmjs.com/package/@fluentui/svg-icons), using this command.
```
[the remark this answers is not in the document]
Comment (dvoituron) [reply src/Icons/Generator/ReadMe.md:10]: Yes. The command line should work correctly with the repo also. I will check that and I will update this page

File: src/Icons/FluentUI.Emojis/README.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
+# Microsoft Fluent UI Emoji's for Blazor
```
[the remark this answers is not in the document]
Comment (dvoituron) [reply src/Icons/FluentUI.Emojis/README.md:1]: Both are available: https://dictionary.cambridge.org/dictionary/english/emoji

File: src/Microsoft.Fast.Components.FluentUI/Components/Icons/CoreIcons.cs (hunk 1 above)
Comment (dvoituron) [reply src/Microsoft.Fast.Components.FluentUI/Components/Icons/CoreIcons.cs:11]: Yes: here to be independent of the optional Icon lib (but `internal`) and in the public Icon lib.

**Comment by mrpmorris** [comment]
Is there an issue number of this that explains the problem + proposed solution?

**Pull request merged**